blueman/gui/manager/ManagerMenu.py

```diff
@@ -132,85 +158,56 @@ def on_device_selected(self, List, device, tree_iter):
         else:
             self.item_device.props.sensitive = False
 
-    def on_adapter_property_changed(self, lst, adapter, kv):
-        (key, value) = kv
-        if key == "Name" or key == "Alias":
-            self.generate_adapter_menu()
-        elif key == "Discovering":
+    def on_adapter_property_changed(self, _adapter, name, value, path):
+        if name == "Name" or name == "Alias":
+            item = self.adapter_items[path][0]
+            item.set_label(value)
+        elif name == "Discovering":
             if self.Search:
                 if value:
                     self.Search.props.sensitive = False
                 else:
                     self.Search.props.sensitive = True
 
-    def generate_adapter_menu(self):
-        menu = Gtk.Menu()
+    def on_adapter_selected(self, menuitem, adapter_path):
+        if menuitem.props.active:
+            if adapter_path != self.blueman.List.Adapter.get_object_path():
+                dprint("selected", adapter_path)
+                self.blueman.Config["last-adapter"] = adapter_path_to_name(adapter_path)
+                self.blueman.List.SetAdapter(adapter_path)
 
-        sep = Gtk.SeparatorMenuItem()
-        sep.show()
-        menu.append(sep)
+    def on_adapter_added(self, _manager, adapter_path):
+        adapter = bluez.Adapter(adapter_path)
+        menu = self.item_adapter.get_submenu()
+        object_path = adapter.get_object_path()
 
-        settings = create_menuitem("_Preferences", get_icon("preferences-system", 16))
-        settings.connect("activate", lambda x: self.blueman.adapter_properties())
-        settings.show()
-        menu.append(settings)
+        item = Gtk.RadioMenuItem.new_with_label(self._adapters_group, adapter['Alias'])
+        item.show()
+        self._adapters_group = item.get_group()
 
-        group = []
-        for adapter in self.adapters:
-            item = Gtk.RadioMenuItem.new_with_label(group, adapter.get_name())
-            group = item.get_group()
+        item_sig = item.connect("activate", self.on_adapter_selected, object_path)
+        adapter_sig = adapter.connect_signal("property-changed", self.on_adapter_property_changed)
 
-            item.connect("activate", self.on_adapter_selected, adapter.get_object_path())
-            if adapter.get_object_path() == self.blueman.List.Adapter.get_object_path():
-                item.props.active = True
+        menu.insert(item, self._insert_adapter_item_pos)
+        self._insert_adapter_item_pos += 1
 
-            item.show()
-            menu.prepend(item)
+        self.adapter_items[object_path] = (item, item_sig, adapter, adapter_sig)
 
-        sep = Gtk.SeparatorMenuItem()
-        sep.show()
-        menu.prepend(sep)
+        if adapter_path == self.blueman.List.Adapter.get_object_path():
+            item.props.active = True
```

Comment:
I looked at this before changing and the one actually calling ``SetAdapter`` (which emits the signal) on the List is this same menuitem through ``on_adapter_selected`` :smile:. Also we now handle adapter-added/removed locally and set the select the active adapter accordingly here.